Shopify/cli
# Add dev xp preview for JS functions

### WHY are these changes introduced?

Implement JS function support in the CLI

### WHAT is this pull request doing?

- Add a new env variable to enable JavaScript templates for functions: `SHOPIFY_CLI_FUNCTIONS_JAVASCRIPT`
- Add new secondary namespace "app function"
- Add a "app function build" command with a default implementation for JS function where we:
  - run `graphql-code-generator` to generate graphQL types
  - run `esbuild` to bundle the function JS
  - run `javy` to compile the JS into wasm
- Add a "app function typegen" command to only generate graphQL types
- Add a "app function run" command to be able to run functions locally

### How to test your changes?

- Create a new app `pnpm create-app --local --path ~/Desktop`
- cd into the newly created app
- Run `export SHOPIFY_CLI_FUNCTIONS_JAVASCRIPT=1 pnpm generate extension --type product_discounts --clone-url https://github.com/Shopify/function-examples-javascript-ju`
- Choose JavaScript or TypeScript
- Cd into the newly generated extension
- Open the folder in VSCode
- Verify type inference is working
- Run `pnpm build`, should succeed
- Run `pnpm typegen`, should succeed
- Run `echo '{ "discountNode": { "metafield": null } }' | pnpm preview`, should succeed

### Post-release steps
- update doc for `generate schema` which now has be to run from the function folder (https://github.com/Shopify/shopify-dev/pull/30870)

### Measuring impact

How do we know this change was effective? Please choose one:


### Checklist
